presto-main/src/main/java/com/facebook/presto/tracing/testing/TestingTracer.java

```diff
@@ -0,0 +1,94 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tracing.testing;
+
+import com.facebook.presto.spi.tracing.Tracer;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class TestingTracer
+        implements Tracer
+{
+    public final Map<String, TestingTracerBlock> blockMap = new ConcurrentHashMap<>();
+    public final Map<String, TestingTracerBlock> recorderBlockMap = new LinkedHashMap<>();
+    public final List<TestingTracerPoint> pointList = new ArrayList<>();
+
+    public TestingTracer()
+    {
+        addPoint("Start tracing");
+    }
+
+    @Override
+    public synchronized void addPoint(String annotation)
+    {
+        pointList.add(new TestingTracerPoint(annotation));
+    }
+
+    @Override
+    public void startBlock(String blockName, String annotation)
+    {
+        if (blockMap.containsKey(blockName)) {
+            throw new RuntimeException("Duplicated block inserted: " + blockName);
+        }
+        TestingTracerBlock block = new TestingTracerBlock(annotation);
+        blockMap.put(blockName, block);
+        synchronized (recorderBlockMap) {
+            recorderBlockMap.put(blockName, block);
+        }
+    }
+
+    @Override
+    public void addPointToBlock(String blockName, String annotation)
+    {
+        if (!blockMap.containsKey(blockName)) {
+            throw new RuntimeException("Adding point to non-existing block: " + blockName);
+        }
+        blockMap.get(blockName).addPoint(new TestingTracerPoint(annotation));
+    }
+
+    @Override
+    public void endBlock(String blockName, String annotation)
+    {
+        if (!blockMap.containsKey(blockName)) {
+            throw new RuntimeException("Trying to end a non-existing block: " + blockName);
+        }
+        blockMap.remove(blockName);
+        synchronized (recorderBlockMap) {
+            recorderBlockMap.get(blockName).end(annotation);
+        }
+    }
+
+    @Override
+    public void endTrace(String annotation)
+    {
+        synchronized (pointList) {
+            pointList.add(new TestingTracerPoint(annotation));
+        }
+    }
+
+    private String getTracingDetails()
+    {
+        return new StringBuilder()
+                .append("{\n")
+                .append("\"blocks\":")
+                .append(recorderBlockMap.toString().replace('=', ':')).append(",")
+                .append("\"points\":")
+                .append(pointList.toString())
```

Comment:
this will only print the memory address of `pointList`

Reply:
It will actually print the list content as I have implemented toString in SimpleTracerPoint. But I'll change to google's helper to make it cleaner